### PR TITLE
Homework003

### PR DESCRIPTION
**Pull request by rodkevich**

### CODE REVIEW

File: homework/hw003/arrayhw/solution-hw-003-arrays.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+// Input ...
+type Input = [6]float64
+
+// Average ...
+type Average func() float64
+
+// ArrayWithFunction ...
+type ArrayWithFunction struct {
```
Comment (DzmitryYafremenka) [inline homework/hw003/arrayhw/solution-hw-003-arrays.go:15]: It's bad style to mix behavior and data.
Structure fields are for data. 
Methods are for behavior.
We can embed types that have their own methods.  But adding a function there is a very bad style.

Besides - check what will happen if you write the following code
```
	var a ArrayWithFunction
	fmt.Println(a.Avg())
```

File: homework/hw003/arrayhw/solution-hw-003-arrays.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"fmt"
+	"strconv"
+)
+
+// Input ...
+type Input = [6]float64
+
+// Average ...
+type Average func() float64
```
Comment (DzmitryYafremenka) [inline homework/hw003/arrayhw/solution-hw-003-arrays.go:12]: Why do we need this?

File: homework/hw003/arrayhw/solution-hw-003-arrays.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+// ArrayWithFunction ...
+type ArrayWithFunction struct {
+	Input *Input
+	Avg   Average // to try adding function in place
+}
+
+// PrintDetailedAverageForArrImp try to add function to type in a pkg
+func (arr *ArrayWithFunction) PrintDetailedAverageForArrImp(res float64) {
```
Comment (DzmitryYafremenka) [inline homework/hw003/arrayhw/solution-hw-003-arrays.go:21]: This function just "prettify value when printing". How is it connected to ArrayWithFunction type?

File: homework/hw003/slicehw/solution-hw-003-slices.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	// process depending of sliceOfStrings length
+	switch len(sliceOfStrings) {
+	case 0:
+		return 0, "", err
+	case 1:
+		return len(sliceOfStrings[0]), sliceOfStrings[0], err
+	default:
+		// skip smaller values
+		for _, value := range sliceOfStrings {
+			if len(value) < res.len {
```
Comment (DzmitryYafremenka) [inline homework/hw003/slicehw/solution-hw-003-slices.go:28]: try 从 as one of the inputs

**Review by DzmitryYafremenka — CHANGES_REQUESTED** [review]

File: cmd/hw003/solution-hw-003.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	// не знаю как отвечать на заданный вопрос почему input ограничен до [6]float64  =)
+	// по заданию нужно работать с array a не slice
+
+	arr, l, avg := arrayhw.CountAverageOfArray(input)
+	value := strconv.FormatFloat(avg, 'f', -1, 64)
+
+	fmt.Printf("arrayshw | result | array=%v len=%v, Avg()=%v", arr, l, value)
+
+	// PART 2: Slices - 1 (longest string in slice):
+	var sliceCases = []struct{ input []string }{
```
Comment (DzmitryYafremenka) [inline cmd/hw003/solution-hw-003.go:27]: Why do we need struct? You only keep one type of data there( []string ) if you need array of inputs you can just use slice of []string -> [][]string.

File: cmd/hw003/solution-hw-003.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		{[]string{"one", "two", "three"}},
+		{[]string{"one", "two"}},
+	}
+	for _, each := range sliceCases {
+		l, max, err := slicehw.LongestStrInSlice(each.input)
```
Comment (DzmitryYafremenka) [inline cmd/hw003/solution-hw-003.go:32]: each.input - You definitely don't need structure here. You just passing one field as function arg.

File: homework/hw003/slicehw/solution-hw-003-slices.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
+package slicehw
+
+import (
+	"errors"
+	"unicode/utf8"
+)
+
+type results struct {
```
Comment (DzmitryYafremenka) [inline homework/hw003/slicehw/solution-hw-003-slices.go:8]: Why do we need this? We only need one string value as result.

File: homework/hw003/slicehw/solution-hw-003-slices.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+		for _, element := range sliceOfStrings {
+			// skip smaller values
+			if length(element) < result.length {
+				continue
+			}
+			// clear smaller values in resulting obj
+			if length(element) > result.length {
+				result.length = length(element)
+				result.list = result.list[:0]
```
Comment (DzmitryYafremenka) [inline homework/hw003/slicehw/solution-hw-003-slices.go:38]: How is this result.list works? What should it contain? And why do we need it?

File: homework/hw003/slicehw/solution-hw-003-slices.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+		return result.length, result.list[0], err
+	}
+}
+
+// ReverseSliceOfInts returns a reversed slice of ints
+func ReverseSliceOfInts(intSlice []int64) []int64 {
+	var rtn []int64
+	ind := len(intSlice) // find length of slice which is max index
+
+	for range intSlice {
```
Comment (DzmitryYafremenka) [inline homework/hw003/slicehw/solution-hw-003-slices.go:52]: Why do you mix `for  `and `for range` semantics? in this case you should use `for`.

**Review by DzmitryYafremenka — CHANGES_REQUESTED** [review]

File: cmd/hw003/solution-hw-003.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
+package main
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/rodkevich/go-course/homework/hw003/arrayhw"
+	"github.com/rodkevich/go-course/homework/hw003/maphw"
+	"github.com/rodkevich/go-course/homework/hw003/slicehw"
```
Comment (IgorZhadenovEpam) [inline cmd/hw003/solution-hw-003.go:9]: Do we realy need 3 pckages? One is more than enough

File: cmd/hw003/solution-hw-003.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+func main() {
+
+	// PART 1: Arrays (average):
+	var input = []float64{1, 2, 3, 4, 5, 6}
+
+	// переделал на преобразование в array внутри функции потому что
+	// не знаю как отвечать на заданный вопрос почему input ограничен до [6]float64  =)
+	// по заданию нужно работать с array a не slice
```
Comment (IgorZhadenovEpam) [inline cmd/hw003/solution-hw-003.go:19]: heh
[]float64{1, 2, 3, 4, 5, 6} - it is not an array :) 
[6]float64{1, 2, 3, 4, 5, 6} - this is an array
I can't see any array in CountAverageOfArray also
read https://blog.golang.org/slices-intro
If you need to convert array to slice:
```
var input = [6]float64{1, 2, 3, 4, 5, 6}
arr, l, avg := arrayhw.CountAverageOfArray(input[:])
```

File: cmd/hw003/solution-hw-003.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+	arr, l, avg := arrayhw.CountAverageOfArray(input)
```
Comment (IgorZhadenovEpam) [inline cmd/hw003/solution-hw-003.go:21]: It is nice that you know about multiple return values but we need just one (average). Don't overcomplicate your code unnecessarily

File: homework/hw003/slicehw/solution-hw-003-slices.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	"unicode/utf8"
+)
+
+type results struct {
+	list   []string
+	length int // set length to be 0
+}
+
+// LongestStrInSlice returns the longest word from the slice of strings
+func LongestStrInSlice(sliceOfStrings []string) (length int, s string, err error) {
```
Comment (IgorZhadenovEpam) [inline homework/hw003/slicehw/solution-hw-003-slices.go:14]: You are playing with code but result is unreadable and overcomplicated. Is it possible to get the longest word using 5-7 lines of code?

File: homework/hw003/arrayhw/solution-hw-003-arrays.go [hunk 14]
```diff
@@ -0,0 +1,25 @@
+package arrayhw
+
+import (
+	"fmt"
+	"strconv"
+)
+
+// Input ...
+type Input = [6]float64
```
[the remark this answers is not in the document]
Comment (rodkevich) [reply homework/hw003/arrayhw/solution-hw-003-arrays.go]: done

**Review by IgorZhadenovEpam — APPROVED** [review]

File: cmd/hw003/solution-hw-003.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+		sum        float64
+		out        float64
+		inputArray = arrayhw.Input{1, 2, 3, 4, 5, 6}
+	)
+
+	var awf = arrayhw.ArrayWithFunction{
+
+		Input: &inputArray,
+		// implement function to count average value of inputArray
+		Avg: func() float64 {
```
[the remark this answers is not in the document]
Comment (rodkevich) [reply cmd/hw003/solution-hw-003.go]: fixed

**Pull request merged**